From a85b23aed7f6081c97b1bfe06eb6aaddc40b878d Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Fri, 10 Jun 2011 17:44:50 -0600
Subject: [PATCH] Allow comment editing with single click on line numbers

Touchscreen based browsers which prevent double clicking
on web pages (android devices zoom on double click) make
it difficult or impossible to comment on patches.  Make
it easier and possible to comment on a patch by simply
clicking on the line number.

Change-Id: Ifb800b15113eab65bd7b7c5c4a618d92b3ec67e4
---
 .../client/patches/AbstractPatchContentTable.java  |    6 +++++-
 .../gerrit/client/patches/SideBySideTable.java     |    9 +++++++++
 .../gerrit/client/patches/UnifiedDiffTable.java    |    9 +++++++++
 3 files changed, 23 insertions(+), 1 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/AbstractPatchContentTable.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/AbstractPatchContentTable.java
index d5a5367..20589de 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/AbstractPatchContentTable.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/AbstractPatchContentTable.java
@@ -331,9 +331,12 @@ public abstract class AbstractPatchContentTable extends NavigationTable<Object>
     return getRowItem(row) instanceof CommentList;
   }
 
-  /** Invoked when the user clicks on a table cell. */
+  /** Invoked when the user double clicks on a table cell. */
   protected abstract void onCellDoubleClick(int row, int column);
 
+  /** Invoked when the user clicks on a table cell. */
+  protected abstract void onCellSingleClick(int row, int column);
+
   /**
    * Invokes createCommentEditor() with an empty string as value for the comment
    * parent UUID. This method is invoked by callers that want to create an
@@ -620,6 +623,7 @@ public abstract class AbstractPatchContentTable extends NavigationTable<Object>
           final int row = rowOf(td);
           if (getRowItem(row) != null) {
             movePointerTo(row);
+            onCellSingleClick(rowOf(td), columnOf(td));
             return;
           }
           break;
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/SideBySideTable.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/SideBySideTable.java
index 1c27388..77f1bf1 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/SideBySideTable.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/SideBySideTable.java
@@ -64,6 +64,13 @@ public class SideBySideTable extends AbstractPatchContentTable {
   }
 
   @Override
+  protected void onCellSingleClick(final int row, int column) {
+    if (column == 1 || column == 3) {
+      onCellDoubleClick(row, column);
+    }
+  }
+
+  @Override
   protected void onInsertComment(final PatchLine line) {
     final int row = getCurrentRow();
     createCommentEditor(row + 1, 4, line.getLineB(), (short) 1);
@@ -335,7 +342,9 @@ public class SideBySideTable extends AbstractPatchContentTable {
       final boolean fullBlock) {
     m.openTd();
     m.setStyleName(Gerrit.RESOURCES.css().lineNumber());
+    m.openAnchor();
     m.append(lineNumberMinusOne + 1);
+    m.closeAnchor();
     m.closeTd();
 
     m.openTd();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/UnifiedDiffTable.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/UnifiedDiffTable.java
index 21a5211..8a8ef45 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/UnifiedDiffTable.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/UnifiedDiffTable.java
@@ -65,6 +65,13 @@ public class UnifiedDiffTable extends AbstractPatchContentTable {
   }
 
   @Override
+  protected void onCellSingleClick(final int row, final int column) {
+    if (column == 2) {
+      onCellDoubleClick(row, column);
+    }
+  }
+
+  @Override
   protected void onInsertComment(final PatchLine pl) {
     final int row = getCurrentRow();
     switch (pl.getType()) {
@@ -374,7 +381,9 @@ public class UnifiedDiffTable extends AbstractPatchContentTable {
   private void appendLineNumber(final SafeHtmlBuilder m, final int idx) {
     m.openTd();
     m.setStyleName(Gerrit.RESOURCES.css().lineNumber());
+    m.openAnchor();
     m.append(idx + 1);
+    m.closeAnchor();
     m.closeTd();
   }
 }
-- 
1.7.4.1

